pki/blacklist/blacklist.go

```diff
@@ -0,0 +1,241 @@
+/*
+ * Nuts node
+ * Copyright (C) 2023 Nuts community
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+package blacklist
+
+import (
+	"crypto/x509"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"sync/atomic"
+	"time"
+
+	config "github.com/nuts-foundation/nuts-node/pki/blacklist/config"
+
+	"github.com/lestrrat-go/jwx/jwa"
+	"github.com/lestrrat-go/jwx/jwk"
+	"github.com/lestrrat-go/jwx/jws"
+)
+
+var (
+	// ErrBlacklistMissing occurs when the blacklist cannot be downloaded
+	ErrBlacklistMissing = errors.New("blacklist cannot be retrieved")
+
+	// ErrCertBlacklisted means the certificate was blacklisted rather than revoked by a CRL
+	ErrCertBlacklisted = errors.New("certificate is blacklisted")
+)
+
+type Blacklist interface {
+	LastUpdated() time.Time
+	Update() error
+	URL() string
+	ValidateCert(cert *x509.Certificate) error
+}
+
+// blacklistImpl implements arbitrary certificate rejection using issuer and serial number tuples
+type blacklistImpl struct {
+	// url specifies the URL where the blacklist is downloaded
+	url string
+
+	// entries is the decoded entries from the downloaded blacklist
+	entries atomic.Pointer[[]blacklistEntry]
+
+	// trustedKey is an Ed25519 key which must sign the blacklist
+	trustedKey jwk.Key
+
+	// lastUpdated contains the time the certificate was last updated
+	lastUpdated time.Time
+}
+
+// blacklistEntry contains parameters for an X.509 certificate that must not be accepted for TLS connections
+type blacklistEntry struct {
+	// Issuer is a string representation (x509.Certificate.Issuer.String()) of the certificate
+	Issuer string
+
+	// SerialNumber is a string representation (x509.Certificate.SerialNumber.String()) of the certificate
+	SerialNumber string
+
+	// JWKThumbprint is an identifier of the public key per https://www.rfc-editor.org/rfc/rfc7638
+	JWKThumbprint string
```

Comment:
why?

Reply:
Because Issuer+SerialNumber isn't enough to uniquely identify a certificate. Ask Gerard for more details